### ISSUE TITLE
precompiler2 fails to compile

### ISSUE THREAD

**Issue by eseidel** [dart-lang/sdk#54860]
Not really sure why.  `dart compile` compiles it fine:

main.dart
```dart
import 'a.dart';

void main() {
  print(a);
}
```

a.dart:
```dart
import 'b.dart';

const Function() a = b;
```

b.dart:
```dart
void b() {}
```

```
 %     DART_CONFIGURATION=DebugARM64 pkg/vm/tool/precompiler2 crash/main.dart main.aot

../../runtime/vm/object.cc: 19656: error: expected: !unit.IsNull()
version=3.2.6 (stable) (Wed Jan 24 13:41:58 2024 +0000) on "macos_arm64"
pid=11064, thread=259, isolate_group=isolate(0x15500be00), isolate=(nil)(0x0)
os=macos, arch=arm64, comp=no, sim=no
isolate_instructions=0, vm_instructions=0
fp=16ef16fd0, sp=16ef16fa8, pc=1012ed854
  pc 0x00000001012ed854 fp 0x000000016ef16fd0 dart::Profiler::DumpStackTrace(bool)+0x28
  pc 0x00000001012ed7bc fp 0x000000016ef17020 dart::Profiler::DumpStackTrace(void*)+0x30
  pc 0x00000001019fabb4 fp 0x000000016ef17040 Dart_DumpNativeStackTrace+0x18
  pc 0x0000000100f5eb78 fp 0x000000016ef17070 dart::Assert::Fail(char const*, ...) const+0x40
  pc 0x000000010123600c fp 0x000000016ef17150 dart::LoadingUnit::LoadingUnitOf(dart::Function const&)+0xf4
  pc 0x000000010179be24 fp 0x000000016ef17190 dart::FlowGraphCompiler::CanPcRelativeCall(dart::Code const&) const+0x5c
  pc 0x00000001017a0b64 fp 0x000000016ef17200 dart::FlowGraphCompiler::ArchSpecificInitialization()+0x64
  pc 0x000000010178b170 fp 0x000000016ef173a0 dart::FlowGraphCompiler::FlowGraphCompiler(dart::compiler::Assembler*, dart::FlowGraph*, dart::ParsedFunction const&, bool, dart::SpeculativeInliningPolicy*, dart::GrowableArray<dart::Function const*> const&, dart::GrowableArray<dart::TokenPosition> const&, dart::GrowableArray<long> const&, dart::ZoneGrowableArray<dart::ICData const*>*, dart::CodeStatistics*)+0x518
  pc 0x000000010178b4e0 fp 0x000000016ef17430 dart::FlowGraphCompiler::FlowGraphCompiler(dart::compiler::Assembler*, dart::FlowGraph*, dart::ParsedFunction const&, bool, dart::SpeculativeInliningPolicy*, dart::GrowableArray<dart::Function const*> const&, dart::GrowableArray<dart::TokenPosition> const&, dart::GrowableArray<long> const&, dart::ZoneGrowableArray<dart::ICData const*>*, dart::CodeStatistics*)+0x94
  pc 0x00000001017179f8 fp 0x000000016ef17ec0 dart::PrecompileParsedFunctionHelper::Compile(dart::CompilationPipeline*)+0x6b4
  pc 0x0000000101718778 fp 0x000000016ef19020 dart::PrecompileFunctionHelper(dart::Precompiler*, dart::CompilationPipeline*, dart::Function const&, bool)+0x380
  pc 0x0000000101711798 fp 0x000000016ef19180 dart::Precompiler::CompileFunction(dart::Precompiler*, dart::Thread*, dart::Zone*, dart::Function const&)+0x174
  pc 0x00000001017106e4 fp 0x000000016ef192b0 dart::Precompiler::ProcessFunction(dart::Function const&)+0x238
  pc 0x000000010170bf70 fp 0x000000016ef19310 dart::Precompiler::Iterate()+0xac
  pc 0x0000000101707b30 fp 0x000000016ef1a1b0 dart::Precompiler::DoCompileAll()+0x14a0
  pc 0x00000001017065f4 fp 0x000000016ef1a650 dart::Precompiler::CompileAll()+0x64
  pc 0x00000001019f7d0c fp 0x000000016ef1a840 Dart_Precompile+0x230
  pc 0x0000000100f169d0 fp 0x000000016ef1a930 dart::bin::CreateAndWritePrecompiledSnapshot()+0x5c
  pc 0x0000000100f13798 fp 0x000000016ef1aa40 dart::bin::CreateIsolateAndSnapshot(dart::bin::CommandLineOptions const&)+0x3b4
  pc 0x0000000100f12b0c fp 0x000000016ef1abf0 dart::bin::main(int, char**)+0x448
  pc 0x0000000100f13920 fp 0x000000016ef1ac10 main+0x24
  pc 0x00000001861ad0e0 fp 0x000000016ef1ae40 start+0x938
-- End of DumpStackTrace
=== Crash occurred when compiling file:///Users/eseidel/Documents/GitHub/dart-sdk/sdk/crash/b.dart_::_b_b in AOT mode in unknown pass
=== Flow Graph not available
pkg/vm/tool/precompiler2: line 121: 11064 Abort trap: 6           "$BIN_DIR"/gen_snapshot "$GEN_SNAPSHOT_OPTION" "$GEN_SNAPSHOT_FILENAME" "${OPTIONS[@]}" "$SNAPSHOT_FILE.dill"
```

**Comment by eseidel** [comment]
This was using 3b128c5454834a1aaef37d9bb12595e7c217ab61.  I've not yet tried main, will do that now.

**Comment by eseidel** [comment]
I was using this for my local testing while reducing:
```sh
#!/bin/bash -x -e

./tools/build.py -m debug -a arm64 --no-goma --gn-args=dart_debug_optimization_level=0 gen_snapshot vm_platform_strong.dill dart_precompiled_runtime --export-compile-commands
DART_CONFIGURATION=DebugARM64 pkg/vm/tool/precompiler2 crash/main.dart main.aot
```

**Comment by eseidel** [comment]
It doesn't crash on main. 😭   So something must just be broken in 3.2.6.

```
eseidel@erics-mbp sdk % ./test_crash.sh
+ ./tools/build.py -m debug -a arm64 --no-goma --gn-args=dart_debug_optimization_level=0 gen_snapshot vm_platform_strong.dill dart_precompiled_runtime --export-compile-commands
Generating Xcode projects took 33ms
Generating compile_commands took 23ms
Done. Made 435 targets from 123 files in 1309ms
buildtools/ninja/ninja -C xcodebuild/DebugARM64 gen_snapshot vm_platform_strong.dill dart_precompiled_runtime
ninja: Entering directory `xcodebuild/DebugARM64'
[1775/1775] LINK ./gen_snapshot
The build took 113.729 seconds
+ DART_CONFIGURATION=DebugARM64
+ pkg/vm/tool/precompiler2 crash/main.dart main.aot

**Comment by eseidel** [comment]
Looks like this is https://github.com/dart-lang/sdk/issues/54324.

**Comment by a-siva** [comment]
> Looks like this is #54324.

Would you need a cherry pick of that fix into 3.2.6 ?

**Comment by eseidel** [comment]
Nah. We build our own Dart and can cherry-pick it ourselves. I'm just confused why this isn't crashing for upstream flutter. It doesn't crash on this assert in release, but it does seem to crash (at least for our dart).

**Comment by eseidel** [comment]
For context, it turns out this was what was causing crashes for Shorebird users this last week. 😭 

https://github.com/shorebirdtech/shorebird/issues/1698#issuecomment-1935281189

I suspect this also causes problems in upstream Flutter, they just probably don't result in a crash.  I've not yet tried to modify the example to make it crash in upstream Flutter, but I suspect there are observable results.  I think it ends up returning out of the same function twice instead of once, but I'm not sure, I've not yet traced it.

**Comment by eseidel** [comment]
Hmm, I'm now seeing after my cherry-pick:

```
../../third_party/dart/runtime/vm/object.cc: 19676: error: Unable to find loading unit of package:dio/src/compute/compute_io.dart_::_compute_compute (class Library:'package:dio/src/compute/compute_io.dart' Class: ::, library Library:'package:dio/src/compute/compute_io.dart')
version=3.2.6 (stable) (Thu Feb 8 19:50:06 2024 -0800) on "macos_arm"
pid=1448, thread=-1, isolate_group=isolate(0x7fed84028600), isolate=(nil)(0x0)
os=macos, arch=arm, comp=no, sim=no
isolate_instructions=0, vm_instructions=0
fp=30cff4ec0, sp=30cff4e90, pc=104644e5b
  pc 0x0000000104644e5b fp 0x000000030cff4ec0 Dart_DumpNativeStackTrace+0x4b
  pc 0x000000010415d961 fp 0x000000030cff4fa0 dart::Assert::Fail(char const*, ...) const+0x91
  pc 0x00000001042dc897 fp 0x000000030cff4ff0 dart::LoadingUnit::LoadingUnitOf(dart::Function const&)+0x117
  pc 0x000000010443b6f5 fp 0x000000030cff5090 dart::FlowGraphCompiler::FlowGraphCompiler(dart::compiler::Assembler*, dart::FlowGraph*, dart::ParsedFunction const&, bool, dart::SpeculativeInliningPolicy*, dart::GrowableArray<dart::Function const*> const&, dart::GrowableArray<dart::TokenPosition> const&, dart::GrowableArray<long> const&, dart::ZoneGrowableArray<dart::ICData const*>*, dart::CodeStatistics*)+0xa85
  pc 0x0000000104407dc2 fp 0x000000030cff5780 dart::PrecompileParsedFunctionHelper::Compile(dart::CompilationPipeline*)+0xd42
  pc 0x0000000104406d92 fp 0x000000030cff5e10 dart::PrecompileFunctionHelper(dart::Precompiler*, dart::CompilationPipeline*, dart::Function const&, bool)+0x5d2
  pc 0x00000001044051e9 fp 0x000000030cff5f20 dart::Precompiler::CompileFunction(dart::Precompiler*, dart::Thread*, dart::Zone*, dart::Function const&)+0x1e9
  pc 0x00000001043ed724 fp 0x000000030cff6c60 dart::Precompiler::DoCompileAll()+0xa154
  pc 0x00000001043e3159 fp 0x000000030cff70e0 dart::Precompiler::CompileAll()+0x989
  pc 0x000000010463fb93 fp 0x000000030cff7250 Dart_Precompile+0x2f3
  pc 0x0000000104156335 fp 0x000000030cff7790 dart::bin::main(int, char**)+0x1795
  pc 0x0000000204da5386 fp 0x000000030cff79b0 Unknown symbol
-- End of DumpStackTrace
=== Crash occurred when compiling package:dio/src/compute/compute_io.dart_::_compute_compute in AOT mode in unknown pass
=== Flow Graph not available
Dart snapshot generator failed with exit code -6
../../third_party/dart/runtime/vm/object.cc: 19676: error: Unable to find loading unit of package:dio/src/compute/compute_io.dart_::_compute_compute (class Library:'package:dio/src/compute/compute_io.dart' Class: ::, library Library:'package:dio/src/compute/compute_io.dart')
version=3.2.6 (stable) (Thu Feb 8 19:50:06 2024 -0800) on "macos_simarm64"
pid=1449, thread=-1, isolate_group=isolate(0x7fa891816e00), isolate=(nil)(0x0)
os=macos, arch=arm64, comp=yes, sim=yes
isolate_instructions=0, vm_instructions=0
fp=30d0a6f00, sp=30d0a6ed0, pc=1048865db
  pc 0x00000001048865db fp 0x000000030d0a6f00 Dart_DumpNativeStackTrace+0x4b
  pc 0x0000000104370d51 fp 0x000000030d0a6fe0 dart::Assert::Fail(char const*, ...) const+0x91
  pc 0x00000001044fd34e fp 0x000000030d0a7030 dart::LoadingUnit::LoadingUnitOf(dart::Function const&)+0x15e
  pc 0x0000000104677aaa fp 0x000000030d0a70d0 dart::FlowGraphCompiler::FlowGraphCompiler(dart::compiler::Assembler*, dart::FlowGraph*, dart::ParsedFunction const&, bool, dart::SpeculativeInliningPolicy*, dart::GrowableArray<dart::Function const*> const&, dart::GrowableArray<dart::TokenPosition> const&, dart::GrowableArray<long> const&, dart::ZoneGrowableArray<dart::ICData const*>*, dart::CodeStatistics*)+0xa8a
  pc 0x0000000104643c7c fp 0x000000030d0a77c0 dart::PrecompileParsedFunctionHelper::Compile(dart::CompilationPipeline*)+0xc3c
  pc 0x0000000104642d02 fp 0x000000030d0a7e50 dart::PrecompileFunctionHelper(dart::Precompiler*, dart::CompilationPipeline*, dart::Function const&, bool)+0x5d2
  pc 0x0000000104640e59 fp 0x000000030d0a7f60 dart::Precompiler::CompileFunction(dart::Precompiler*, dart::Thread*, dart::Zone*, dart::Function const&)+0x1e9
  pc 0x0000000104627c65 fp 0x000000030d0a8c90 dart::Precompiler::DoCompileAll()+0xa475
  pc 0x000000010461d379 fp 0x000000030d0a9120 dart::Precompiler::CompileAll()+0xc29
  pc 0x0000000104881333 fp 0x000000030d0a9290 Dart_Precompile+0x2f3
  pc 0x00000001043693b5 fp 0x000000030d0a97d0 dart::bin::main(int, char**)+0x1815
  pc 0x0000000204c44386 fp 0x000000030d0a99f0 Unknown symbol
-- End of DumpStackTrace
=== Crash occurred when compiling package:dio/src/compute/compute_io.dart_::_compute_compute in AOT mode in unknown pass
=== Flow Graph not available
Dart snapshot generator failed with exit code -6
../../third_party/dart/runtime/vm/object.cc: 19676: error: Unable to find loading unit of package:dio/src/compute/compute_io.dart_::_compute_compute (class Library:'package:dio/src/compute/compute_io.dart' Class: ::, library Library:'package:dio/src/compute/compute_io.dart')
version=3.2.6 (stable) (Thu Feb 8 19:50:06 2024 -0800) on "macos_x64"
pid=1450, thread=-1, isolate_group=isolate(0x7f8f21828600), isolate=(nil)(0x0)
os=macos, arch=x64, comp=yes, sim=no
isolate_instructions=0, vm_instructions=0
fp=30df9bf10, sp=30df9bee0, pc=104e9b7db
  pc 0x0000000104e9b7db fp 0x000000030df9bf10 Dart_DumpNativeStackTrace+0x4b
  pc 0x0000000104989d51 fp 0x000000030df9bff0 dart::Assert::Fail(char const*, ...) const+0x91
  pc 0x0000000104b159fe fp 0x000000030df9c040 dart::LoadingUnit::LoadingUnitOf(dart::Function const&)+0x15e
  pc 0x0000000104c874ca fp 0x000000030df9c0e0 dart::FlowGraphCompiler::FlowGraphCompiler(dart::compiler::Assembler*, dart::FlowGraph*, dart::ParsedFunction const&, bool, dart::SpeculativeInliningPolicy*, dart::GrowableArray<dart::Function const*> const&, dart::GrowableArray<dart::TokenPosition> const&, dart::GrowableArray<long> const&, dart::ZoneGrowableArray<dart::ICData const*>*, dart::CodeStatistics*)+0xa8a
  pc 0x0000000104c4d7ec fp 0x000000030df9c7d0 dart::PrecompileParsedFunctionHelper::Compile(dart::CompilationPipeline*)+0xc3c
  pc 0x0000000104c4c872 fp 0x000000030df9ce60 dart::PrecompileFunctionHelper(dart::Precompiler*, dart::CompilationPipeline*, dart::Function const&, bool)+0x5d2
  pc 0x0000000104c4a9c9 fp 0x000000030df9cf70 dart::Precompiler::CompileFunction(dart::Precompiler*, dart::Thread*, dart::Zone*, dart::Function const&)+0x1e9
  pc 0x0000000104c317a5 fp 0x000000030df9dca0 dart::Precompiler::DoCompileAll()+0xa475
  pc 0x0000000104c26eb9 fp 0x000000030df9e130 dart::Precompiler::CompileAll()+0xc29
  pc 0x0000000104e96533 fp 0x000000030df9e2a0 Dart_Precompile+0x2f3
  pc 0x00000001049823b5 fp 0x000000030df9e7e0 dart::bin::main(int, char**)+0x1815
  pc 0x0000000205520386 fp 0x000000030df9ea00 Unknown symbol
-- End of DumpStackTrace
=== Crash occurred when compiling package:dio/src/compute/compute_io.dart_::_compute_compute in AOT mode in unknown pass
=== Flow Graph not available
Dart snapshot generator failed with exit code -6
Target android_aot_release_android-arm failed: Exception: AOT snapshotter exited with code -6
Target android_aot_release_android-arm64 failed: Exception: AOT snapshotter exited with code -6
Target android_aot_release_android-x64 failed: Exception: AOT snapshotter exited with code -6
```

I suspect my cherry pick was incomplete.

**Comment by eseidel** [comment]
Actually my cherry-pick is complete, it's just that this fix is in the .dill generation, which we currently use stock dart for.  This would be our first time modifying that side of Dart.  So either we need to patch shorebird to use our modified Dart (which means we need to make sure our modified Dart still works correctly for JIT, etc) or we need to use a different version of Dart (e.g. beta) with current stable flutter to get this fix. 🤔

**Comment by eseidel** [comment]
I think we're going to give up on trying to fix 3.16.9 with Shorebird and instead start to prepare for 3.19.0.  We would need to do a bunch of work to fix this ourselves, or request a CP, but my sense is 3.19.0 is very soon now, so seem silly to ask/wait for a CP.

**Comment by eseidel** [comment]
Looks like Dart 3.3.0 just shipped, so we'll just move to that: https://github.com/dart-lang/sdk/commit/bbb92d0b75d436d15c0d0fb53b81588241e7207d
